pkg/ride/functions_proto.go

```diff
@@ -15,178 +15,104 @@ import (
 		}
-
 		intAmount, ok := amount.(rideInt)
 		if !ok {
-			return nil, errors.Errorf("reentrantInvoke: unexpected argument type '%s'", amount.instanceOf())
+			return nil, RuntimeError.Errorf("property 'amount' of attached payment %d has an invalid type '%s'",
+				i, amount.instanceOf())
 		}
-		var asset *proto.OptionalAsset
-
-		switch asID := assetID.(type) {
-		case rideBytes:
-			asset, err = proto.NewOptionalAssetFromBytes(asID)
-			if err != nil {
-				return nil, errors.Errorf("reentrantInvoke: failed to get optional asset from ride bytes")
-			}
-		case rideUnit:
-			waves := proto.NewOptionalAssetWaves()
-			asset = &waves
-		default:
-			return nil, errors.Errorf("reentrantInvoke: unexpected argument type '%s'", args[0].instanceOf())
+		assetID, err := payment.get("assetId")
+		if err != nil {
+			return nil, RuntimeError.Wrap(err, "attached payment")
 		}
-
-		attachedPayments = append(attachedPayments, proto.ScriptPayment{Asset: *asset, Amount: uint64(intAmount)})
-	}
-	// since RideV5 the limit of attached payments is 10
-	if len(attachedPayments) > 10 {
-		return nil, errors.New("reentrantInvoke: no more than ten payments is allowed since RideV5 activation")
-	}
-
-	var attachedPaymentActions []proto.ScriptAction
-	for _, payment := range attachedPayments {
-		payment := &proto.AttachedPaymentScriptAction{Sender: &callerPublicKey, Recipient: recipient, Amount: int64(payment.Amount), Asset: payment.Asset}
-		attachedPaymentActions = append(attachedPaymentActions, payment)
-	}
-
-	address, err := env.state().NewestRecipientToAddress(recipient)
-	if err != nil {
-		return nil, errors.Errorf("reentrantInvoke: failed to get address from dApp, invokeFunctionFromDApp")
-	}
-	env.setNewDAppAddress(*address)
-	err = ws.smartAppendActions(attachedPaymentActions, env)
-	if err != nil {
-		return nil, errors.Wrapf(err, "reentrantInvoke: failed to apply attached payments")
-	}
-
-	if ws.invCount() > 1 {
-		if isAddressInBL(*recipient.Address, ws.blackList) && proto.WavesAddress(callerAddress) != *recipient.Address {
-			return rideUnit{}, errors.Errorf("function call of %s with dApp address %s is forbiden because it had already been called once by 'invoke'", fnName, recipient.Address)
+		asset, err := extractOptionalAsset(assetID)
+		if err != nil {
+			return nil, RuntimeError.Errorf("property 'assetId' of attached payment %d has an invalid type '%s': %v",
+				i, assetID.instanceOf(), err)
 		}
+		res[i] = proto.ScriptPayment{Asset: asset, Amount: uint64(intAmount)}
 	}
+	return res, nil
+}
 
-	res, err := invokeFunctionFromDApp(env, recipient, fnName, listArg)
-	if err != nil {
-		return nil, errors.Wrapf(err, "reentrantInvoke")
-	}
-
-	if res.Result() {
-		if res.UserError() != "" {
-			return nil, errors.Errorf(res.UserError())
+func extractFunctionName(v rideType) (rideString, error) {
+	switch tv := v.(type) {
+	case rideUnit:
+		return "default", nil
+	case rideString:
+		if tv == "" {
+			return "default", nil
 		}
```

Comment:
Are you sure that returning `default` is right in both cases?

Reply:
Yes, `unit` or empty string means the default function, so we return "default"